### ISSUE TITLE
have global time budget for tunin

### ISSUE THREAD

**Issue by berndbischl** [mlr-org/mlr#1239]
like: 
"you can tune this problem with mlr, you have 8h in total for everything"

**Comment by larskotthoff** [comment]
How exactly should this work? Check after each evaluation if enough time is left for more, or potentially disrupt evaluations?

**Comment by berndbischl** [comment]
> How exactly should this work? Check after each evaluation if enough time is left for more, or potentially disrupt evaluations?

one possible route is this: 
check after each eval, whether the cutoff time is violated. if so, we need to make the tuner stop, but we dont have control over it. so there are only 2 options. 
a) send an exception,. this is not so cool as the tuner can catch these. also the tuner will then not reach a valid state and return a result object (how important that is,  i dont know, as we might take the best config from the opt.path anyway). 
b) for all remaining evals simply return a very bad y value (basically the "y-impute" value from the TuneControl)

I think i like b) better. Remaining problems are:

* The cutoff time is not perfectly respected. if the last eval is very costly that can be a problem. but i currently dont see a better (simple) way.
* one needs to define what the cutoff means if we eval in parallel. i would suggest to simply consider the "clock" on the master process. that seems most doable, and also realistic.

**Comment by larskotthoff** [comment]
I don't think that this is a good idea. This assumes that the tuner won't do a lot of expensive things with bad y values, which isn't necessarily the case. So the cutoff value could be exceeded by an almost arbitrarily large amount that is outside of our control.

I think the only way of doing this properly would be to run the tuner (and the evaluations) in a separate R process. Yes, that's a huge pain, but that's how Auto-WEKA and Auto-sklearn are doing this to get reasonably close the timeouts the user specifies.

To do something quick and useful, it may be a better idea to keep track of the evaluations done and based on that show a "progress bar" or similar that shows an estimate of how long the tuning will take. Then the user can at least abort manually and restart with adjusted parameters if it looks like it would take too long.

**Comment by berndbischl** [comment]
> To do something quick and useful, it may be a better idea to keep track of the evaluations done and based on that show a "progress bar" or similar that shows an estimate of how long the tuning will take. Then the user can at least abort manually and restart with adjusted parameters if it looks like it would take too long.

that's something we should do for some mlr operations in general, like resampling and other stuff. (and i think we discussed this already someplace else). it is a different issue than what is discussed here.

regarding the case you made:

1) running the target evals in a separate process has many advantages. but: we already support that! with the batchjobs / parallelmap integration! that also gives you walltimes when you are on a cluster.
i really (?) dont want to do more here. i dont even know how that would be supported on windows? and even on linux this might be a hassle to do.

2) it does not even help with the point you are raising? you do realize that?

> this assumes that the tuner won't do a lot of expensive things with bad y values, which isn't necessarily the case.

can you name an example where the tuner "does bad expensive things" still? i can name 1. 
bayesian optimization. here the tuner itself does expensive stuff in between objective function eval. all other tuners are pretty cheap in between.
now how does running the eval in a separate R process help here? it does not. 

you basically have to tell the tuner the time budget it has been given, so the tuner can stop itself. which means you need that option in EVERY tuner interface (which you dont control, and this does not exist for most tuners we support).

but mlrMBO actually (as we have written it) *does* support that time budget options. so we *can* simply pass it down. and for the other tuners implement what i outlined above?

**Comment by larskotthoff** [comment]
I mean run *both* the evaluation and the tuner in separate processes.

**Comment by berndbischl** [comment]
> I mean run both the evaluation and the tuner in separate processes.

ok, got it now. 
ok then: if i tell you now to program this for mlr, lets make it easy and only say for linux, can you program this? so it runs on my ubuntu laptop?

**Comment by berndbischl** [comment]
(i am not ordering you to do it, i just want to find out if we even know how this could be programmed)

**Comment by larskotthoff** [comment]
Not off the top of my head. I don't think even prototyping this will be that easy. We need to share various objects etc between the processes in a robust way (i.e. when something is killed we want to have at least partial data).

**Comment by berndbischl** [comment]
its not that i havent thought about this.....
see here:
https://github.com/berndbischl/BBmisc/blob/master/todo-files/callWithTimeout.R

i also discussed some other options with @mb706 when he worked on automlr (same usecase as discussed here...) but i cannot remember everything he googled then. there was at least an r package to do this, but i am not sure how robust and useful it was.
